### PR TITLE
JS: Sanitizer for `sanitizer(x) === true`

### LINKED ISSUE #11667: False positive: About Javascript TaintBarriers
I give some tests about Javascript TaintBarriers library. These show when SanitizerGuardNode compares with true/false or inverts will make things wrong.  Details show as below.

I custom a Sanitizer called `CustomAdHocWhitelistCheckSanitizer` in `javascript/ql/test/library-tests/TaintBarriers/ExampleConfiguration.qll`, and make `outcome` includes  `[true,false]` in order to observe SINK points in true/false branch.

```java
class CustomAdHocWhitelistCheckSanitizer extends TaintTracking::AdHocWhitelistCheckSanitizer {
  override predicate sanitizes(boolean outcome, Expr e) {
    outcome = [true, false] and
    e = this.getArgument(0).asExpr()
  }
}
```

### First case

```javascript
1  function conditionSanitization() {
2     var v = SOURCE();
3
4     if (isWhitelisted(v)) {
5        SINK(v);
6     } else {
7        SINK(v);
8     }
9  }
```

Result as expected.

```log
+isBarrier
+isLabeledBarrier
+| ExampleConfiguration | tst.js:5:14:5:14 | v | taint |
+| ExampleConfiguration | tst.js:7:14:7:14 | v | taint |
+isSanitizer
+sanitizingGuard
+| tst.js:4:9:4:24 | isWhitelisted(v) | tst.js:4:23:4:23 | v | false |
+| tst.js:4:9:4:24 | isWhitelisted(v) | tst.js:4:23:4:23 | v | true |
+taintedSink
```

### Second case

```javascript
1  function conditionSanitization() {
2     var v = SOURCE();
3
4     if (isWhitelisted(v) === true) {
5        SINK(v);
6     } else {
7        SINK(v);
8     }
9  }
```

something wrong. SINK in line 5 and 7 are all labeled as `taintedSink`.

```log
+isBarrier
+isLabeledBarrier
+isSanitizer
+sanitizingGuard
+| tst.js:4:9:4:24 | isWhitelisted(v) | tst.js:4:23:4:23 | v | false |
+| tst.js:4:9:4:24 | isWhitelisted(v) | tst.js:4:23:4:23 | v | true |
+| tst.js:4:9:4:33 | isWhite ... == true | tst.js:4:9:4:24 | isWhitelisted(v) | true |
+| tst.js:4:9:4:33 | isWhite ... == true | tst.js:4:30:4:33 | true | true |
+taintedSink
+| tst.js:2:13:2:20 | SOURCE() | tst.js:5:14:5:14 | v |
+| tst.js:2:13:2:20 | SOURCE() | tst.js:7:14:7:14 | v |
```

### Third case

```javascript
1  function conditionSanitization() {
2     var v = SOURCE();
3
4     if (~isWhitelisted(v)) {
5        SINK(v);
6     } else {
7        SINK(v);
8     }
9  }
```

wrong. SINK in line 5 and 7 are all labeled as `taintedSink`.

```log
+isBarrier
+isLabeledBarrier
+isSanitizer
+sanitizingGuard
+| tst.js:4:10:4:25 | isWhitelisted(v) | tst.js:4:24:4:24 | v | false |
+| tst.js:4:10:4:25 | isWhitelisted(v) | tst.js:4:24:4:24 | v | true |
+taintedSink
+| tst.js:2:13:2:20 | SOURCE() | tst.js:5:14:5:14 | v |
+| tst.js:2:13:2:20 | SOURCE() | tst.js:7:14:7:14 | v |
```

### PR DESCRIPTION
**Pull request by erik-krogh**
Fixes #11667

The `appliesTo` change didn't change the result of any tests, but that feels like a lucky coincidence (because the nodes were `instanceof DataFlow::BarrierGuardNode` anyway).  

Evaluations were uneventful:   
- [nightly on code-scanning with meta queries](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/moreSan__nightly-old__code-scanning__1/reports)
- [default on code-scanning](https://github.com/github/codeql-dca-main/tree/data/erik-krogh/moreSan__default__code-scanning__2/reports)

### CODE REVIEW

File: javascript/ql/lib/semmle/javascript/Expr.qll [hunk 1]
```diff
@@ -379,7 +379,10 @@ class NullLiteral extends @null_literal, Literal { }
  * false
  * ```
  */
-class BooleanLiteral extends @boolean_literal, Literal { }
+class BooleanLiteral extends @boolean_literal, Literal {
+  /** Gets the value of this literal. */
+  boolean getBoolValue() { if this.getRawValue() = "true" then result = true else result = false }
```
Comment (esbena) [inline javascript/ql/lib/semmle/javascript/Expr.qll:384]: Maybe consider naming consistency with TypeInference.qll:

```ql
  /** Gets the unique Boolean value that this node evaluates to, if any. */
  boolean getTheBooleanValue() { ... }
  ```

Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/Expr.qll:384]: `getBoolValue` is consistent with the predicate in `YamlBool` (In `YAML.qll`).   
So I prefer to keep it as is.

File: javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll [hunk 2]
```diff
@@ -2002,6 +2004,42 @@ private class AdditionalBarrierGuardCall extends AdditionalBarrierGuardNode, Dat
...
+private class CallAgainstEqualityCheck extends AdditionalBarrierGuardNode {
+  DataFlow::BarrierGuardNode prev;
+  boolean polarity;
+
+  CallAgainstEqualityCheck() {
+    prev instanceof DataFlow::CallNode and
+    exists(EqualityTest test, BooleanLiteral bool |
+      this.asExpr() = test and
+      test.hasOperands(prev.asExpr(), bool) and
+      polarity = test.getPolarity().booleanXor(bool.getBoolValue())
```
Comment (aibaars) [inline javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll:2025]: I'm not sure I understand the use of `xor` here.

**Comment by aibaars** [comment]
There are some merge conflicts.

File: javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/lib/semmle/javascript/dataflow/Configuration.qll:2022]: I need `polarity` to flip both if `test` flips and if `bool` flips.  
`xor` always flips when one of it's inputs flip, that's why I'm using it.  

Then I had to figure out whether I had to add a `.booleanNot()` to the output or not, and that was just done by testing.

**Review by asgerf — APPROVED** [review]
👍

**Pull request merged**